zerver/actions/message_send.py

```diff
@@ -1528,6 +1529,37 @@ def validate_stream_id_with_pm_notification(
     return stream
 
 
+def check_can_send_private_message(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile], recipient: Recipient
+) -> None:  # nocoverage
+    if realm.name == "Realm Test" or sender.is_bot:
+        return
+
+    for user_profile in user_profiles:
+        if not (user_profile.is_bot or user_profile.id == sender.id):
+            break
+    else:
+        return
+
+    allowed_user_group = realm.direct_message_permission_group
+    allowed_user_ids = get_user_group_member_ids(allowed_user_group)
+    if sender.id not in allowed_user_ids:
+        for user_profile in user_profiles:
+            if user_profile.id in allowed_user_ids:
+                break
+        else:
+            raise JsonableError(
+                _("You are not allowed to send direct messages to these recipient(s).")
+            )
+
+    if not sender.has_permission("direct_message_initiator_group"):
+        previous_messages_exist = Message.objects.filter(
+            sender=sender, recipient=recipient, realm=realm
+        ).exists()
+        if not previous_messages_exist:
+            raise JsonableError(_("You are not allowed to initiate direct message thread."))
```

Comment:
You can likely improve performance and use of database queries significantly by using some of the techniques we use in `user_access_restricted_in_realm` and the functions around it. 

This last check is probably wrong for 1:1 DMs -- the other user could have sent you the DM.

Reply:
I have tested it in case of 1:1 DMs and it is working fine. Can you please elaborate as to why do you think it is wrong?

Reply:
See https://github.com/zulip/zulip/issues/25713 for backend, but 1:1 DMs are represented with a sender/recipient pair, and both pairs together make up a single DM conversation.

Reply:
I have a question regarding the design of this setting. Can members who can't initiate DM threads send messages to groups that have messages that were sent before the setting change? (if the previous value for this setting was `role:everyone`).